### PR TITLE
Introduce new ARXML traversal functions

### PR DESCRIPTION
**Pull request by andlaus**
This PR continues the quest for ARXML-3 support, cf https://github.com/Daimler/cantools/tree/arxml_refactor

In this PR a set improved of traversal functions for ARXML documents is introduced and the existing code is changed to use them. The main advantage of them is that they make it very easy to follow ARXML references and to specifies colletions of XML nodes which might have ARXML references in their location.

Be aware that I'm not sure where exactly references are allowed by the AUTOSAR standard or how to obtain that information, so I've enabled them quite liberally. The rationale is that if AUTOSAR does not allow references in a given context but we support them anyway, well-formed files will still load fine.

Andreas Lauser <andreas.lauser@mbition.io>, Mercedes-Benz AG on behalf of [MBition GmbH](https://mbition.io/).

[Imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md#mbition-gmbh)

### CODE REVIEW

**Review by koppor — APPROVED** [review]

File: cantools/database/can/formats/arxml.py [hunk 1]
```diff
@@ -19,149 +19,32 @@
...
+                                     [ "AR-PACKAGES",
+                                       "*AR-PACKAGE",
+                                       "ELEMENTS",
+                                       "*&CAN-CLUSTER",
+                                       "CAN-CLUSTER-VARIANTS",
+                                       "*&CAN-CLUSTER-CONDITIONAL",
+                                       "PHYSICAL-CHANNELS",
+                                       "*&CAN-PHYSICAL-CHANNEL",
+                                       "FRAME-TRIGGERINGS",
+                                       "*&CAN-FRAME-TRIGGERING" ])
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:47]: Replace with the following to match the code style of the project.

```python
can_frame_triggerings = self.find_arxml_children(
    self.root,
    [
        "AR-PACKAGES",
        "*AR-PACKAGE",
        "ELEMENTS",
        "*&CAN-CLUSTER",
        "CAN-CLUSTER-VARIANTS",
        "*&CAN-CLUSTER-CONDITIONAL",
        "PHYSICAL-CHANNELS",
        "*&CAN-PHYSICAL-CHANNEL",
        "FRAME-TRIGGERINGS",
        "*&CAN-FRAME-TRIGGERING"
    ])
```

File: cantools/database/can/formats/arxml.py [hunk 2]
```diff
@@ -179,35 +62,48 @@ def load_message(self, can_frame_triggering):
...
-            time_period = i_signal_i_pdu.find(TIME_PERIOD_XPATH, NAMESPACES)
+        # For "sane" bus systems like CAN or LIN, there ought to be
+        # only a single PDU per frame. AUTOSAR also supports "insane"
+        # bus systems like flexray, though...
+        pdu = self.find_unique_arxml_child(can_frame,
+                                          [
+                                              "PDU-TO-FRAME-MAPPINGS",
+                                              "&PDU-TO-FRAME-MAPPING",
+                                              "&PDU"
+                                          ])
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:88]: Fix indentation.

```python
pdu = self.find_unique_arxml_child(can_frame,
                                   [
                                       "PDU-TO-FRAME-MAPPINGS",
                                       "&PDU-TO-FRAME-MAPPING",
                                       "&PDU"
                                   ])
```

File: cantools/database/can/formats/arxml.py [hunk 3]
```diff
@@ -179,35 +62,48 @@ def load_message(self, can_frame_triggering):
...
+
+        if pdu is not None:
+            time_period = self.find_unique_arxml_child(pdu,
+                                                      [ "I-PDU-TIMING-SPECIFICATIONS",
+                                                        "I-PDU-TIMING",
+                                                        "TRANSMISSION-MODE-DECLARATION",
+                                                        "TRANSMISSION-MODE-TRUE-TIMING",
+                                                        "CYCLIC-TIMING",
+                                                        "TIME-PERIOD",
+                                                        "VALUE" ])
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:98]: Styling:

```python
time_period = self.find_unique_arxml_child(pdu,
                                           [
                                               "I-PDU-TIMING-SPECIFICATIONS",
                                               "I-PDU-TIMING",
                                               "TRANSMISSION-MODE-DECLARATION",
                                               "TRANSMISSION-MODE-TRUE-TIMING",
                                               "CYCLIC-TIMING",
                                               "TIME-PERIOD",
                                               "VALUE"
                                           ])
```

File: cantools/database/can/formats/arxml.py [hunk 4]
```diff
@@ -179,35 +62,48 @@ def load_message(self, can_frame_triggering):
...
             if time_period is not None:
                 cycle_time = int(float(time_period.text) * 1000)
 
-            i_signal_to_i_pdu_mappings = i_signal_i_pdu.iterfind(
-                I_SIGNAL_TO_I_PDU_MAPPING_XPATH,
-                NAMESPACES)
+            i_signal_to_i_pdu_mappings = \
+                self.find_arxml_children(pdu,
+                                         [ "I-SIGNAL-TO-PDU-MAPPINGS",
+                                           "*&I-SIGNAL-TO-I-PDU-MAPPING" ])
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:106]: Styling:

```python
i_signal_to_i_pdu_mappings = self.find_arxml_children(
    pdu,
    [
        "I-SIGNAL-TO-PDU-MAPPINGS",
        "*&I-SIGNAL-TO-I-PDU-MAPPING"
    ])
```

File: cantools/database/can/formats/arxml.py [hunk 5]
```diff
@@ -227,25 +123,10 @@ def load_message(self, can_frame_triggering):
...
-
-    def load_message_length(self, can_frame):
-        return int(can_frame.find(FRAME_LENGTH_XPATH, NAMESPACES).text)
-
-    def load_message_is_extended_frame(self, can_frame_triggering):
-        can_addressing_mode = can_frame_triggering.find(
-            CAN_ADDRESSING_MODE_XPATH,
-            NAMESPACES).text
-
-        return can_addressing_mode == 'EXTENDED'
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:245]: Re-add all these methods to make the code easier to read where used.

File: cantools/database/can/formats/arxml.py [hunk 6]
```diff
@@ -268,37 +149,26 @@ def load_signal(self, i_signal_to_i_pdu_mapping):
...
         decimal = SignalDecimal(Decimal(factor), Decimal(offset))
 
-        i_signal_ref = i_signal_to_i_pdu_mapping.find(
-            I_SIGNAL_REF_XPATH,
-            NAMESPACES)
-
-        if i_signal_ref is None:
-            # Probably a signal group (I-SIGNAL-GROUP-REF).
+        i_signal = self.find_unique_arxml_child(i_signal_to_i_pdu_mapping, "&I-SIGNAL")
+        if i_signal is None:
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:153]: Add a blank line before the `if` statement.

File: cantools/database/can/formats/arxml.py [hunk 7]
```diff
@@ -321,37 +191,28 @@ def load_signal(self, i_signal_to_i_pdu_mapping):
...
-                                              NAMESPACES).text
-
-    def load_signal_start_position(self, i_signal_to_i_pdu_mapping):
-        return int(i_signal_to_i_pdu_mapping.find(
-            START_POSITION_XPATH,
-            NAMESPACES).text)
-
-    def load_signal_length(self, i_signal):
-        return int(i_signal.find(LENGTH_XPATH, NAMESPACES).text)
-
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:335]: Re-add all these methods to make the code easier to read where used.

File: cantools/database/can/formats/arxml.py [hunk 8]
```diff
@@ -321,37 +191,28 @@ def load_signal(self, i_signal_to_i_pdu_mapping):
...
             return 'little_endian'
 
     def load_signal_unit(self, system_signal):
-        unit_ref = system_signal.find(UNIT_REF_XPATH, NAMESPACES)
+        result = self.find_unique_arxml_child(system_signal,
+                                              [ "PHYSICAL-PROPS",
+                                                "SW-DATA-DEF-PROPS-VARIANTS",
+                                                "&SW-DATA-DEF-PROPS-CONDITIONAL",
+                                                "&UNIT",
+                                                "DISPLAY-NAME"])
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:208]: Fix styling.

File: cantools/database/can/formats/arxml.py [hunk 9]
```diff
@@ -466,38 +329,35 @@ def load_system_signal(self, system_signal, decimal):
...
+                    'CATEGORY in compu method {} does not exist.'.format(
+                        compu_method.find("SHORT-NAME").text))
 
             category = category.text
 
             if category == 'TEXTTABLE':
-                minimum, maximum, choices = self.load_texttable(
-                    compu_method,
-                    decimal)
+                minimum, maximum, choices = self.load_texttable(compu_method, decimal)
```
Comment (eerimoq) [inline cantools/database/can/formats/arxml.py:350]: Try not to make styling changes. Both makes the diff bigger and the styling is there for a reason. In this case to make lines shorter than about 85 character. This particular line is 86, which would often be acceptable, but let's not change it now.

**Review by eerimoq — COMMENTED** [review]
Hi!

Looks good, just a bunch of styling comments. I did not write comments for all styling issues. Try to apply similar styling fixes throughout the code.

Please use `'` and not `"` for strings. `"` can be used if the string itself contains a `'`.

Thanks!

**Comment by andlaus** [comment]
Thanks for the review, I included the requested changes. I hope that I did not forget too much, because I was a bit "lost in git branches" for a while.

I also have a few questions/remarks on why the original patches looked like they did:

- It seems that the retrieval of sub-nodes is not done consistently in master: sometimes a `get_*()` method is used, sometimes it seems to be `find_*()` or `load_*()` and sometimes the `ElementTree` is traversed directly
- I normally prefer to avoid the opacity of relatively trivial functions (e.g., getter functions for internal data), in particular if these functions are called from exactly one location.
- I'm probably too much of a C++ guy and have cobbled together too many shell scripts in my live, so I usually use normal quotes for strings ;). since the current master is somewhat inconsistent on this front as well (E.g. the `EcuLoader` class uses double quotes in places which you do not mention above), I thought it would be a good idea unify the style of strings to something that's compatible with most other languages.

**Comment by eerimoq** [comment]
Ok =)

Are there more PR:s after this one? I'll fix the styling myself when everything has been merged as I think that's the quickest way forward. I'm too picky when it comes to styling :P

**Comment by andlaus** [comment]
> Are there more PR:s after this one?

yes, quite a few. I indent get what's in https://github.com/Daimler/cantools/tree/arxml_refactor fully upstream. This PR is basically just "ground clearance" work for the ARXML3 stuff...

> I'll fix the styling myself when everything has been merged as I think that's the quickest way forward. I'm too picky when it comes to styling :P

Let's do it as you prefer, i'm fine if the whole process takes a while as long as the end result is good. I generally don't really care too much about coding style, so I will try to adapt for the remaining PRs, but I am prone to make mistakes, so please be a bit lenient with me ;)

Small question: Shouldn't  internal methods be prefixed by an underscore in python?

**Comment by eerimoq** [comment]
Yeah, they probably should be prefixed with an underscore. In this case the user should never use this module directly as it is not documented in the official documentation. Any functions not part of the documentation is considered internal in this project. I guess that's not clear to anyone but me, but that's why it looks like it does.

**Comment by andlaus** [comment]
ok, I'll adapt the SystemLoader class in the next PR, then...

**Comment by eerimoq** [comment]
Let's just continue merging. It would take forever to fix them all, and I can easily fix it once we are done.

What's important now is that there are tests in place to ensure that the code works, and that future changes wont creak legacy.

**Comment by andlaus** [comment]
I just noticed that the "goal branch" on the Daimler github organization is not public for whatever reason. (Oops!) For your benefit, I thus also uploaded it to my personal account:  https://github.com/andlaus/cantools/tree/arxml_refactor

**Comment by eerimoq** [comment]
Ok, thanks. I'll have I look if I find the time, which is unlikely as I'm focusing on the [Mys project](https://github.com/eerimoq/mys) atm.

**Pull request merged**